### DISCUSSION TITLE
Media Upload with s3 disk and Laravel 9 Error

### DISCUSSION THREAD

**Discussion by discoverlance-com** [spatie/laravel-medialibrary#3151] [Q&A]
I am using
PHP v8.1 and Laravel 9

Whilst trying to upload a media with the disk set to 's3', I always get the error, `Disk named `s3` cannot be accessed`.

I then created a test route just to test it but I got the same error. Find here the image describing the error.

![image](https://user-images.githubusercontent.com/52635726/211195241-c5690eff-719c-479f-ab67-08ad26d6e55e.png)

I tried adding the custom header with ACL set to *public-read* as suggested in this [issue's comment](https://github.com/spatie/laravel-medialibrary/issues/241#issuecomment-349455094) and also in my media-library config file but I still get the same error.

My s3 filesystem disk config:

```php
's3' => [
            'driver' => 's3',
            'key' => env('AWS_ACCESS_KEY_ID'),
            'secret' => env('AWS_SECRET_ACCESS_KEY'),
            'region' => env('AWS_DEFAULT_REGION'),
            'bucket' => env('AWS_BUCKET'),
            'url' => env('AWS_URL'),
            'endpoint' => env('AWS_ENDPOINT'),
            'use_path_style_endpoint' => env('AWS_USE_PATH_STYLE_ENDPOINT', false),
            'throw' => false,
        ],
```

I also tried setting the `visibility="public"` but it still failed

**Comment by levipeto** [comment]
I had a similar issue start happening and what solved mine was set to
config/livewire.php
'temporary_file_upload' => [ 'disk' => 's3' ...]

^ it was  ['disk' => null  ... ] by default.

**Reply by discoverlance-com** [reply]
I already set my disk also to s3 and the uploads still fail. 

The temporary file upload works though, as I can see the files temporarily  in the livewire-tmp directory in my  s3 bucket but when it comes to uploading the file with the laravel-medialibrary, then it fails.

**Comment by blueminded** [comment]
Hello, Devs.

I'm having the same issue.
<img width="650" alt="image" src="https://user-images.githubusercontent.com/2212245/217989208-361b22cd-2ad7-4458-becc-1bba0adc3b63.png">

I have no clue what's going on. I recently add laravel sail an updated my project to Laravel 9 and I have had the same configuration for it and now suddenly doesn't work. Any has found a solution? Thanks in advance.

**Reply by blueminded** [reply]
Well, 

you are not going to believe it!! after a week trying to fix this

The solution was to set the config vars inside quotes, like:
AWS_ACCESS_KEY_ID="my_access_key"
instead of:
AWS_ACCESS_KEY_ID=my_access_key

for all the vars related to file system.

I hope this helps anyone having this problem. However, not sure why this happens since .env file is supposed to work with strings by default.

**Reply by osbre** [reply]
I can confirm that adding quotes has fixed my issue with Cloudflare R2 uploads. Pure magic.
Thank you so much @blueminded!

**Reply by theaungmyatmoe** [reply]
I am having some issues with this, and I have no clue. I already checked out both the code and the ACL in AWS, but neither is working.

**Comment by oktayla** [comment]
I have same issue. Any idea for solution?

**Comment by mass5** [comment]
I have the same issue, I tried everything. 
If I upload a file to S3 via Laravel everything works fine, but the Media Library does not work. I'm pretty disappointed because I lost so much time with Media Pro, I could have built this myself. 🙈

**Reply by oktayla** [reply]
If you're using AWS S3, then update MEDIA_DISK=s3 and add all AWS_ credentials For AWS s3, make sure your properly setup permission of the bucket with ACL enable by follow this link -> https://stackoverflow.com/a/70603995/2158023

**Reply by mass5** [reply]
hey @oktayla thanks for the tip, the ACL enabled on my AWS bucket did the trick. Much appreciated, I lost sooo much time trying to debug this. I think the Media Library needs clear documentation regarding the bucket settings for temporary uploading. I could not find this info on their website. (I have the paid version, the Pro)

**Comment by peter-tharwat** [comment]
make sure that your `AWS_SECRET_ACCESS_KEY` doesn't have special characters like "+" afer i changed SECRET_ACCESS_KEY it works

**Reply by ahbanavi** [reply]
Thanks, the use of `#`, `$`, and `@` in password was causing the problem for me. Even wrapping the password in `"` did not work. The only solution was to remove and change these characters in my password.

**Comment by maulik1211** [comment]
Hope this can help someone.

If none of solution is working and If you do not want to enable ACL, then you can use pre signed urls for users to download the file.

This can be achieved with TemporaryUrl() method.

example:
`$url = $record->getFirstTemporaryUrl(Carbon::now()->addMinutes(15), 'collection-name');`

**Comment by ismoil-nosr** [comment]
In case if someone still struggles with this error. I used container's name for endpoint like  `AWS_ENDPOINT=http://app_s3:9000` and it turns out that problem was with underscore. After I changed container's name to `apps3` everything worked.

**Reply by theaungmyatmoe** [reply]
52 ~ │ AWS_ENDPOINT=https://s3.${AWS_DEFAULT_REGION}.amazonaws.com/${AWS_BUCKET}/

Don't know exactly but container name is not working 🥲

**Reply by ismoil-nosr** [reply]
Check your bucket's name also, does it have underscore ?

**Reply by theaungmyatmoe** [reply]
I currently using dash (-) instead of underscore (_). I will try that out. Thank you.
